### PR TITLE
Fix whole game potentially breaking when entering editor too early from song select

### LINKED ISSUE #22927: Star Difficulty calculation breaks if user press Edit Beatmap while songs in the Song Selection Menu are loading
### Type

Game behaviour

### Bug description

1. Go to the Song Selection Menu
2. Go back to the main menu
3. Go back to the Song Selection Menu
4. Press "options" and then "edit beatmap" before the song in the song selection menu loads
5. Quit the editor
6. The maps don't show any star difficulties and they can't be played

### Screenshots or videos

https://user-images.githubusercontent.com/5209299/226760017-958efd09-2a9b-457b-a9ee-56ec61da8246.mp4

### Version

2023.305.0

### Logs

[performance-audio.log](https://github.com/ppy/osu/files/11034578/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/11034579/performance-draw.log)
[performance-input.log](https://github.com/ppy/osu/files/11034580/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/11034581/performance-update.log)
[runtime.log](https://github.com/ppy/osu/files/11034582/runtime.log)
[updater.log](https://github.com/ppy/osu/files/11034583/updater.log)
[database.log](https://github.com/ppy/osu/files/11034584/database.log)
[input.log](https://github.com/ppy/osu/files/11034585/input.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/11034586/legacy-ipc.log)
[network.log](https://github.com/ppy/osu/files/11034587/network.log)
[performance.log](https://github.com/ppy/osu/files/11034588/performance.log)

### PR DESCRIPTION
**Pull request by peppy**
Resolves https://github.com/ppy/osu/issues/22927.

TL;DR the editor could get loaded while `DummyRuleset` is active, and then restore that ruleset `OnExiting`, causing a cascading failure which means no beatmap can be diffcalc'd or even played until the game-wide ruleset is switched at least once.

To prevent similar scenarios from happening elsewhere, I've forbidden the usage of `Push` on `SongSelect`. It makes handling pushing when song select is involved a bit more arduous, but I think it's for the best (short of `Push` being `virtual` so we could add protection there).

Of note, this means that operations which would previous leave the game in an unknown state will now *not fire* instead. `FinaliseSelection` has early-abort when state is incorrect. In the future we may want to queue the action and run it when the load completes, but I think that's for a separate day.

### CODE REVIEW

File: osu.Game/OsuGame.cs [hunk 1]
```diff
@@ -655,15 +655,23 @@ public void PresentScore(IScoreInfo score, ScorePresentType presentType = ScoreP
                 Ruleset.Value = databasedScore.ScoreInfo.Ruleset;
                 Beatmap.Value = BeatmapManager.GetWorkingBeatmap(databasedBeatmap);
 
-                switch (presentType)
+                if (screen is SongSelect songSelect)
+                    songSelect.FinaliseSelection(databasedBeatmap, databasedScore.ScoreInfo.Ruleset, customStartAction: push);
+                else
+                    push();
```
Comment (bdach) [inline osu.Game/OsuGame.cs:661]: This thing is making it *very* painful for me to accept this fix. I understand the rationale for it, as well as the intention to eliminate this class of bugs entirely by using `FinaliseSelection()` everywhere, but this sort of thing spreading out to `OsuGame`, combined with the added complexity and another footgun to avoid in `SongSelect` (which already feels impossible to meaningfully touch, given how convoluted the decoupled bindables are and the carousel is) makes it much more difficult to swallow in my view.

Like at this point I would probably accept something much more direct over this, like something along the lines of:

```diff
diff --git a/osu.Game/Screens/Select/SongSelect.cs b/osu.Game/Screens/Select/SongSelect.cs
index c5e914b461..dcdab964eb 100644
--- a/osu.Game/Screens/Select/SongSelect.cs
+++ b/osu.Game/Screens/Select/SongSelect.cs
@@ -383,6 +383,12 @@ protected override IReadOnlyDependencyContainer CreateChildDependencies(IReadOnl
 
         public void Edit(BeatmapInfo? beatmapInfo = null)
         {
+            if (!Carousel.BeatmapSetsLoaded)
+            {
+                Logger.Log($"{nameof(Edit)} aborted as carousel beatmaps are not yet loaded");
+                return;
+            }
+
             if (!AllowEditing)
                 throw new InvalidOperationException($"Attempted to edit when {nameof(AllowEditing)} is disabled");
 ```

Probably would like a few other opinions to sound off on this (cc @ppy/team-client)

**Comment by bdach** [comment]
I will *also* note that from the other direction, editor is not entirely without fault in this scenario, because of this:

https://github.com/ppy/osu/blob/7dd113f20800f6d4c7afc638863fc86435925060/osu.Game/Screens/Edit/Editor.cs#L724-L736

Note that `refetchedBeatmap` is checked if it's a `DummyWorkingBeatmap`, but `Beatmap.Value` in the first place *is not*. So part of this is being triggered by the fact that the editor is refetching dummy beatmaps, turning them into *real* beatmaps in the process, which causes the ruleset change:

```
2023-03-22 19:57:45 [verbose]: Game-wide working beatmap updated to please load a beatmap! - no beatmaps available!
2023-03-22 19:57:45 [verbose]: 📺 OsuScreenStack#141(depth:4) suspended PlaySongSelect#485 (waiting on EditorLoader#402)
2023-03-22 19:57:45 [verbose]: 📺 OsuScreenStack#141(depth:5) loading EditorLoader#402
2023-03-22 19:57:45 [debug]: ClickEvent(Left) handled by BeatmapOptionsButton.
2023-03-22 19:57:45 [debug]: MouseClick handled by BeatmapOptionsButton.
2023-03-22 19:57:45 [verbose]: 📺 OsuScreenStack#141(depth:5) entered EditorLoader#402
2023-03-22 19:57:45 [verbose]: 📺 BackgroundScreenStack#490(depth:3) loading BackgroundScreenBeatmap#455
2023-03-22 19:57:45 [verbose]: 📺 OsuScreenStack#141(depth:5) suspended EditorLoader#402 (waiting on Editor#284)
2023-03-22 19:57:45 [verbose]: 📺 OsuScreenStack#141(depth:6) loading Editor#284
(...)
2023-03-22 19:57:47 [debug]: Pressed (Back) handled by Editor.
2023-03-22 19:57:47 [debug]: KeyDownEvent(Escape, False) handled by GlobalActionContainer.
(...)
2023-03-22 19:57:49 [verbose]: Game-wide working beatmap updated to please load a beatmap! - no beatmaps available!
2023-03-22 19:57:49 [verbose]: Editor providing re-fetched beatmap post edit session
2023-03-22 19:57:49 [verbose]: Game-wide working beatmap updated to please load a beatmap! - no beatmaps available!
2023-03-22 19:57:49 [verbose]: 📺 OsuScreenStack#141(depth:5) exit from Editor#284
2023-03-22 19:57:49 [verbose]: 📺 OsuScreenStack#141(depth:5) resume to EditorLoader#402
2023-03-22 19:57:49 [verbose]: 📺 BackgroundScreenStack#490(depth:2) exit from BackgroundScreenBeatmap#455
2023-03-22 19:57:49 [verbose]: 📺 BackgroundScreenStack#490(depth:2) resume to BackgroundScreenBeatmap#231
2023-03-22 19:57:49 [verbose]: 📺 OsuScreenStack#141(depth:4) exit from EditorLoader#402
2023-03-22 19:57:49 [verbose]: 📺 OsuScreenStack#141(depth:4) resume to PlaySongSelect#485
2023-03-22 19:57:49 [verbose]: Song select working beatmap updated to please load a beatmap! - no beatmaps available!
2023-03-22 19:57:49 [verbose]: decoupled ruleset transferred ("osu!" -> "dummy")
```

It doesn't change the fact that it shouldn't probably be possible to enter the editor at all in this state, but food for thought perhaps.

File: osu.Game/OsuGame.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/OsuGame.cs:661]: That's fine, and was the initial solution I had. But I'm not anywhere close to 100% confident that other issues will arise from not running the `FinaliseSelection` flow when pushing a child screen to song select.
